contracts/Pool.sol

```diff
@@ -0,0 +1,381 @@
+    require(recipient != address(this), "ERC20: cannot transfer to the pool address");
+
+    _accumulateDepositInterest(msg.sender);
+
+    require(_deposited[msg.sender] >= amount, "ERC20: transfer amount exceeds balance");
+
+    // (this is verified in "require" above)
+    unchecked {
+      _deposited[msg.sender] -= amount;
+    }
+
+    _accumulateDepositInterest(recipient);
+    _deposited[recipient] += amount;
+
+    emit Transfer(msg.sender, recipient, amount);
+
+    return true;
+  }
+
+  function allowance(address owner, address spender) external view override returns (uint256) {
+    return _allowed[owner][spender];
+  }
+
+  function approve(address spender, uint256 amount) external override returns (bool) {
+    _accumulateDepositInterest(msg.sender);
+
+    require(_deposited[msg.sender] >= amount, "ERC20: approve amount exceeds balance");
+    _allowed[msg.sender][spender] = amount;
+
+    emit Approval(msg.sender, spender, amount);
+
+    return true;
+  }
+
+  function transferFrom(address sender, address recipient, uint256 amount) external override returns (bool) {
+    require(amount <= _allowed[sender][msg.sender], "Not enough tokens allowed to transfer required amount.");
+    require(recipient != address(0), "ERC20: cannot transfer to the zero address");
+    require(recipient != address(this), "ERC20: cannot transfer to the pool address");
+
+    _accumulateDepositInterest(msg.sender);
+
+    require(amount <= _deposited[sender], "Not enough tokens to transfer required amount.");
+
+    _deposited[sender] -= amount;
+    _allowed[sender][msg.sender] -= amount;
+
+    _accumulateDepositInterest(recipient);
+    _deposited[recipient] += amount;
+
+    emit Transfer(sender, recipient, amount);
+
+    return true;
+  }
+
+
+  /**
+   * Deposits the message value
+   * It updates user deposited balance, total deposited and rates
+  **/
+  function deposit() payable virtual external {
+    _accumulateDepositInterest(msg.sender);
+
+    _mint(msg.sender, msg.value);
+    _updateRates();
+
+    emit Deposit(msg.sender, msg.value, block.timestamp);
+  }
+
+  /**
+   * Withdraws selected amount from the user deposits
+   * @dev _amount the amount to be withdrawn
+  **/
+  function withdraw(uint256 _amount) external nonReentrant {
+    _accumulateDepositInterest(msg.sender);
+
+    _burn(msg.sender, _amount);
+
+    payable(msg.sender).transfer(_amount);
+
+    _updateRates();
+
+    emit Withdrawal(msg.sender, _amount, block.timestamp);
+  }
+
+
+  /**
+   * Borrows the specified amount
+   * It updates user borrowed balance, total borrowed amount and rates
+   * @dev _amount the amount to be borrowed
+  **/
+  function borrow(uint256 _amount) payable external canBorrow nonReentrant {
+    require(address(this).balance >= _amount, "There is not enough funds in the pool to fund the loan.");
+    require(totalSupply() - totalBorrowed() >= _amount, "There is no enough deposit in the pool to fund the loan.");
```

Comment:
Let's put a comment that we're checking the maxUtilisation in canBorrow, so it's clear it won't cause an underflow @piotruda

Reply:
done